set.js

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var util = require('util');
+var assert = require('assert');
+var ListSpec = require('./list').ListSpec;
+
+function SetSpec(valueType, annotations) {
+    // TODO consider annotations for {key: true} vs Set vs [] (default)
+    var self = this;
+    ListSpec.call(self, valueType, annotations);
+    self.mode = annotations && annotations['js.type'] || 'array';
+    assert(self.mode === 'array' || self.mode === 'object', 'js.type annotation must be either "array" or "object"');
+}
+
+util.inherits(SetSpec, ListSpec);
+
+SetSpec.prototype.name = 'set';
+
+SetSpec.prototype.create = function create() {
+    var self = this;
+    if (self.mode === 'array') {
+        return [];
+    } else { // if (self.mode === 'object') {
+        return {};
+    }
```

Comment:
All these methods seem to be doing:

    if (self.mode === 'array') {
        // array-based impl
    } else {
        // set based impl
    }

That suggests they should be two separate classes.